### PR TITLE
Refactor model provisioning (cleaned up version of previous PR)

### PR DESCRIPTION
**Pull request by jbemmel**
- put power module provisioning into separate routine
- add method to enable ports with connectors and LLDP
- add support for sr-2s with xiom

### CODE REVIEW

File: sros/docker/launch.py [hunk 1]
```diff
@@ -35,23 +35,47 @@ def trace(self, message, *args, **kws):
...
 SROS_VARIANTS = {
     "ixr-e-big": {
         "deployment_model": "distributed",
         # control plane (CPM)
-        "max_nics": 34,
+        "max_nics": 34, # 24*10 + 8*25G + 2*100G (with connector)
         "cp": {
             "min_ram": 3,
-            "timos_line": "slot=A chassis=ixr-e card=cpm-ixr-e",
+            "timos_line": "slot=A chassis=ixr-e card=cpm-ixr-e/imm24-sfp++8-sfp28+2-qsfp28",
```
Comment (hellt) [inline sros/docker/launch.py:63]: should the cpm of a control plane vm specify the imm?
copying @LimeHat

**Comment by jbemmel** [comment]
[image: image.png]
based on VSIM install guide

File: sros/docker/launch.py (hunk 1 above)
Comment (LimeHat) [reply sros/docker/launch.py:64]: Not needed AFAIK, I've been using this in the past:
```
               slot=A \
                           chassis=IXR-e \
                           card=cpm-ixr-e
```

File: sros/docker/launch.py [hunk 2]
```diff
@@ -35,23 +34,48 @@ def trace(self, message, *args, **kws):
...
+            **LINE_CARD(
+                chassis="ixr-e",
+                card="cpm-ixr-e/imm24-sfp++8-sfp28+2-qsfp28",
+                card_type="imm24-sfp++8-sfp28+2-qsfp28",
+                mda="m24-sfp++8-sfp28+2-qsfp28",
+            ),
         },
+        "connector": {"type": "c1-100g", "ports": [33, 34]},
+        # /chassis power-supply are automatically provisioned by vSR itself
+        # "power": { 'modules': { 'ac/hv': 3, 'dc': 4 } },
```
Comment (hellt) [inline sros/docker/launch.py:78]: I guess since this is commented out - it can be removed?

File: sros/docker/launch.py [hunk 3]
```diff
@@ -400,6 +451,57 @@ def bootstrap_spin(self):
...
+        modules = power_cfg["modules"]
+        if type(modules) is dict:
+            modules = modules[self.power]  # 3(AC) or 4(DC)
+
+        if self.power == "dc":  # vSIM default
+            power_shelf_type = f"ps-a{modules}-shelf-dc"
+            module_type = "ps-a-dc-6000"
+        else:
+            power_shelf_type = f"ps-b{modules}-shelf-ac/hv"
+            module_type = "ps-b-ac/hv-6000"
```
Comment (hellt) [inline sros/docker/launch.py:468]: the AC power modules are used for which systems, considering that vSIMs only accept DC?

**Comment by jbemmel** [comment]
They're not used, I added them before I realized vSIM was DC-only

**Comment by hellt** [comment]
Lets cut it out then? To not have inaccessible code

**Comment by jbemmel** [comment]
There remains the possibility of having containerlab manage (configurations
for) physical devices some day, but we can remove it (comment it out) until
then

File: sros/docker/launch.py [hunk 4]
```diff
@@ -35,23 +34,47 @@ def trace(self, message, *args, **kws):
...
+    Note that it may be possible to omit configuring these explicitly
+    """
+    slot = "A" if integrated else "1"
+    return {
+        "timos_line": f"slot={slot} chassis={chassis} card={card} mda/1={mda}",
+        "card_config": f"""
+       /configure card 1 card-type {card_type if card_type else card}
+       /configure card 1 mda 1 mda-type {mda}
+      """,
+    }
```
Comment (hellt) [inline sros/docker/launch.py:54]: I am doing some comments to the parts that have been surgically merged in other PRs

**Comment by hellt** [comment]
I've gradually pulled everything from this PR except for port provisioning function.

**Comment by hellt** [comment]
since most (maybe all) of these enhancements have been ported in subsequent PRs this might be closed

**Pull request closed without merging**